How many late days are you using for this assignment? 0

In [ ]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [ ]:
# Load the given datasets
categorical_df = pd.read_csv("chronic_kidney_disease_categorical.csv")
numerical_df = pd.read_csv("chronic_kidney_disease_numerical.csv")

# Print the data
print(categorical_df)
print(numerical_df)


     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: This data is collected from individuals to predict chronic kidney disease and is multivariate. The features are the independent variables and the labels are the dependent variables we want to predict. For example, class tells us whether a patient has ckd or not. 
The chronic_kidney_disease_categorical.csv dataset contains both catergorical and numerical features, while the chronic_kidney_disease_numerical.csv dataset only contains numerical features. Th e numerical features are measurable quantities like age and blood pressure. Categorical features are the descriptions like yes or no. 

3 numerical features are age, bp (blood pressure), and bgr (blood glucose random level)

3 categorical features are htn (hypertension), appet (appetite condition), and rbc (red blood cells)	

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [47]:
import pandas as pd
numerical_df = pd.read_csv("chronic_kidney_disease_numerical.csv")
categorical_df = pd.read_csv("chronic_kidney_disease_categorical.csv")

# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
num_dups = numerical_df.duplicated(subset=['unique_id']).sum()
print(f"Total number of duplicate rows: {num_dups}")
#30 duplicate rows

# Drop these duplicate rows
numerical_df = numerical_df.drop_duplicates(subset=['unique_id'], keep='first')
print(numerical_df)
#[390 rows x 12 columns]

# Repeat the same for categorical dataset. Print the duplicate rows and drop them

cat_dups = categorical_df.duplicated(subset=['unique_id']).sum()
print(f"Total number of duplicate rows: {cat_dups}")
#Total number of duplicate rows: 30

categorical_df = categorical_df.drop_duplicates(subset=['unique_id'], keep='first')
print(categorical_df)
#[400 rows x 14 columns]

Total number of duplicate rows: 30
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
415     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0   
416     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
417     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0   
418     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
419     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0   

        wbcc  rbcc  
0     7300.0   6.4  
1     7200.0   5.5  
2     5500.0 

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: All the unique ids are not present in both datasets and we know that because of the given dimensions as well as no duplicates. It cannot be that the categorical dataset has all of the unique ids of the numerical dataset. In a combined table, the rows missing look like NaN or even exlcuded completely.

In [31]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
num_unique_ids = set(numerical_df['unique_id'])
cat_unique_ids = set(categorical_df['unique_id'])
merged = pd.merge(numerical_df, categorical_df, on="unique_id", how="outer")

#Print the combined dataset
print(merged)
#[400 rows x 25 columns]

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   
397     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0  ...   
398     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0  ...   
399     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [19]:
# Calculate the percentage of rows that contain atleast one missing value
percentage_missed = (merged.isnull().any(axis=1).sum() / len(merged)) * 100

# Print %
print(percentage_missed)
#61.75000000000001

# Drop these rows from the dataset
clean_merged_df = merged.dropna()

# Print the Dataset
print(clean_merged_df)
#[153 rows x 25 columns]

61.75000000000001
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0  ...   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0  ...   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc        

## Part 3.4: Sort the dataset according to the Labels

In [20]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_merged_df = clean_merged_df.sort_values(by="Target").reset_index(drop=True)

# Print the dataset
print(sorted_merged_df)
#[153 rows x 25 columns]

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [21]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encoded_merged_df = pd.get_dummies(sorted_merged_df, columns=categorical_columns, drop_first=False)

# Print the dataset
print(encoded_merged_df)
#[153 rows x 36 columns]

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: In the downstream modeling, this would be equivalent since both methods would provide the same information. But if the categorical variable took more than 2 values, the single column would not be equivalent since it may cause the model to treat categorical variables like numerical ones. In the sandwich example, the single column for the binary variable may incorrectly imply that the condiments are in an order when they should bei independent. 

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [ ]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

def remove_outliers(merged):
    for col in numerical_columns:
        mean = merged[col].mean()
        std_dev = merged[col].std()
        upper = mean + (3 * std_dev)
        lower = mean - (3 * std_dev)
        merged = merged[(merged[col] >= lower) & (merged[col] <= upper)]
    return merged

filtered = remove_outliers(encoded_merged_df)
filtered.reset_index(drop=True, inplace=True)

# Print the dataset

print(filtered)


     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
1       397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0  ...   
2       438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0  ...   
3       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
4       475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
128     486397  20.0   70.0  123.0   44.0  1.0  135.0  3.8  14.6  44.0  ...   
129     203466  22.0   60.0   97.0   18.0  1.2  138.0  4.3  13.5  42.0  ...   
130     482105  44.0   60.0   95.0   46.0  0.5  138.0  4.2  15.0  50.0  ...   
131     228849  58.0   80.0  100.0   50.0  1.2  140.0  3.5  14.0  50.0  ...   
132     996076  57.0   60.0  105.0   49.0  1.2  150.0  4.7  15.7  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [37]:
# Normalize the all Numerical Attributes in the dataset.
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

def remove_outliers(merged):
    for col in numerical_columns:
        mean = merged[col].mean()
        std_dev = merged[col].std()
        upper = mean + (3 * std_dev)
        lower = mean - (3 * std_dev)
        merged = merged[(merged[col] >= lower) & (merged[col] <= upper)]
    return merged

filtered = remove_outliers(encoded_merged_df)
filtered.reset_index(drop=True, inplace=True)

for col in numerical_columns:
    min_val = filtered[col].min()
    max_val = filtered[col].max()
    filtered[col] = (filtered[col] - min_val) / (max_val - min_val)

# Print the dataset
print(filtered)
#[133 rows x 36 columns]

     unique_id       age    bp       bgr        bu        sc       sod  \
0       343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
1       397388  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
2       438182  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
3       474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
4       475200  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
..         ...       ...   ...       ...       ...       ...       ...   
128     486397  0.189189  0.25  0.227468  0.222222  0.074074  0.500000   
129     203466  0.216216  0.00  0.115880  0.052288  0.098765  0.600000   
130     482105  0.513514  0.00  0.107296  0.235294  0.012346  0.600000   
131     228849  0.702703  0.50  0.128755  0.261438  0.098765  0.666667   
132     996076  0.689189  0.00  0.150215  0.254902  0.098765  1.000000   

          pot      hemo       pcv  ...  cad_no  cad_yes  appet_good  \
0    0.793103  0.000000  0.032258  ...  

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The unique_id column should be removed because they are unique identifiers and wouldn't help the model predict anything. This variable doesn't have any relationship in terms of predicting the target value. 

In [ ]:
#Remove that column
filtered.drop(columns=['unique_id'], inplace=True)

# Print the dataset
print(filtered)
#[133 rows x 35 columns]

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [25]:
# Export the dataframe to a new csv file
filtered.to_csv("cleaned_dataset.csv", index=False)

# Export the dataframe to a new json file
filtered.to_json("cleaned_dataset.json", orient="records", lines=True)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

| age  | bp   | bgr  | bu   | sc   | sod  | pot  | hemo | pcv  | wbcc | rbcc | al  | su  | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|------|------|------|------|------|------|------|------|------|------|------|------|------|--------------|------------|-------------|------------|---------------|-------------|---------------|------------|--------|--------|-------|-------|-------|-------|------------|------------|------|------|------|------|------------|--------------|
| 0.743 | 0.5  | 0.442 | 0.902 | 0.432 | 0.5  | 0.793 | 0.0  | 0.032 | 0.395 | 0.057 | 2.0  | 0.0  | True  | False | True  | False | True  | False | True  | False | False | True  | False | True  | False | True  | False | True  | False | True  | False | True  | False |
| 0.770 | 1.0  | 0.901 | 0.163 | 0.346 | 0.767 | 0.207 | 0.525 | 0.548 | 0.444 | 0.343 | 2.0  | 2.0  | False | True  | False | True  | True  | False | False | True  | False | True  | True  | False | False | True  | True  | False | True  | False | True  | False | True  |
| 0.905 | 0.5  | 0.785 | 0.863 | 0.519 | 0.6  | 1.0  | 0.277 | 0.323 | 0.234 | 0.371 | 2.0  | 0.0  | True  | False | True  | False | True  | False | True  | False | False | True  | False | True  | False | True  | True  | False | True  | False | True  | False | True  |
| 0.203 | 0.75 | 0.159 | 0.196 | 0.160 | 0.167 | 0.207 | 0.059 | 0.0  | 0.653 | 0.257 | 4.0  | 0.0  | False | True  | True  | False | False | True  | False | True  | True  | False | True  | False | True  | False | True  | False | True  | False | False | True  | True  |
| 0.784 | 1.0  | 0.399 | 0.288 | 0.840 | 0.667 | 0.586 | 0.020 | 0.097 | 0.258 | 0.114 | 4.0  | 2.0  | True  | False | True  | False | True  | False | False | True  | False | True  | False | True  | True  | False | True  | False | False | True  | True  | False | True  |
| 0.730 | 0.0  | 0.936 | 0.170 | 0.160 | 0.333 | 0.034 | 0.020 | 0.065 | 0.879 | 0.000 | 3.0  | 1.0  | False | True  | True  | False | False | True  | True  | False | False | True  | True  | False | True  | False | False | True  | True  | False | False | True  | True  |
| 0.676 | 0.75 | 0.253 | 0.634 | 0.778 | 0.367 | 0.655 | 0.139 | 0.194 | 0.169 | 0.114 | 2.0  | 0.0  | True  | False | True  | False | True  | False | True  | False | False | True  | True  | False | True  | False | True  | False | True  | False | True  | False | True  |
| 0.851 | 0.25 | 0.618 | 0.562 | 0.728 | 0.0  | 0.345 | 0.168 | 0.161 | 0.581 | 0.086 | 4.0  | 3.0  | False | True  | True  | False | False | True  | False | True  | False | True  | False | True  | False | True  | True  | False | False | True  | False | True  | True  |
| 0.541 | 0.0  | 0.399 | 0.536 | 0.358 | 0.7  | 0.379 | 0.208 | 0.161 | 0.831 | 0.057 | 1.0  | 0.0  | False | True  | False | True  | True  | False | True  | False | False | True  | False | True  | True  | False | True  | False | True  | False | True  | False | True  |
| 0.757 | 0.25 | 0.223 | 0.209 | 0.160 | 0.533 | 0.621 | 0.485 | 0.516 | 0.290 | 0.257 | 3.0  | 0.0  | False | True  | True  | False | True  | False | True  | False | False | True  | False | True  | True  | False | True  | False | True  | False | True  | False | True  |
| 0.662 | 0.5  | 0.618 | 0.412 | 0.432 | 0.567 | 0.690 | 0.317 | 0.355 | 0.25  | 0.200 | 3.0  | 1.0  | False | True  | True  | False | False | True  | False | True  | False | True  | False | True  | True  | False | True  | False | False | True  | True  | False | True  |
| 0.784 | 0.0  | 0.725 | 0.314 | 0.481 | 0.567 | 0.862 | 0.178 | 0.194 | 0.258 | 0.114 | 4.0  | 1.0  | True  | False | True  | False | True  | False | False | True  | False | True  | False | True  | True  | False | False | True  | False | True  | True  | False | True  |
| 0.878 | 0.0  | 0.206 | 0.752 | 0.605 | 0.533 | 0.690 | 0.366 | 0.387 | 0.879 | 0.371 | 4.0  | 0.0  | False | True  | False | True  | True  | False | True  | False | False | True  | False | True  | True  | False | False | True  | False | True  | True  | False | True  |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [40]:
# Code to rename all the columns in the dataset
import pandas as pd

# Sample DataFrame
data = {
    'age': [25, 30],
    'bp': [120, 130],
    'sg': [1.020, 1.025],
    'al': [0, 1],
    'su': [0, 1],
    'rbc': [4.2, 4.5],
    'pc': [0, 2],
    'pcc': [0, 1],
    'ba': [0, 1],
    'bgr': [90, 110],
    'bu': [15, 20],
    'sc': [0.8, 1.0],
    'sod': [135, 140],
    'pot': [4.0, 4.5],
    'hemo': [13.5, 14.0],
    'pcv': [40, 42],
    'wc': [6000, 6500],
    'rc': [4.0, 4.2],
    'htn': [0, 1],
    'dm': [0, 1],
    'cad': [0, 1],
    'appet': [1, 1],
    'pe': [0, 1],
    'ane': [0, 1],
    'class': [0, 1]
}

df = pd.DataFrame(data)

# Mapping of abbreviations to full names
rename_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Rename the DataFrame columns using the mapping
df.rename(columns=rename_mapping, inplace=True)

# Display the updated DataFrame
print(df)


   age  blood pressure  specific gravity  albumin  sugar  red blood cells  \
0   25             120             1.020        0      0              4.2   
1   30             130             1.025        1      1              4.5   

   pus cell  pus cell clumps  bacteria  blood glucose random  ...  \
0         0                0         0                    90  ...   
1         2                1         1                   110  ...   

   packed cell volume  white blood cell count  red blood cell count  \
0                  40                    6000                   4.0   
1                  42                    6500                   4.2   

   hypertension  diabetes mellitus  coronary artery disease  appetite  \
0             0                  0                        0         1   
1             1                  1                        1         1   

   pedal edema  anemia  class  
0            0       0      0  
1            1       1      1  

[2 rows x 25 columns]


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [6]:
# Converted SQL to Pandas code

import pandas as pd
df = pd.read_csv('cleaned_dataset.csv')  

query = df.groupby('Target_ckd').size().reset_index(name='count')
print(query)


   Target_ckd  count
0       False    111
1        True     22


*Describe what the above code does here*

This code selects the Target columns and counts the number of rows from the cleaned dataset and displays them. The group by line groups the output by the number of occurrences of each target. 